### PR TITLE
Add a "version" key for the training data files

### LINKED ISSUE #6129: New training data format (Maturation)
Umbrella ticket for the maturation of the New Training Data Format.

We can logically split the work into parts:

Feedback on `2.0`:

- [x] NLU: intents should start with `-` as noted [here](https://www.notion.so/rasa/Feedback-on-2-0-4574eaacc0764967ac0b8b82fbc72904?d=092eb214-51e1-4125-8148-bcfa1eb4e2af#6929b68580314eb59140ff23032df35b) - #6243 

Interactive learning:
- [x] NLU: needs new YAML writer - #6252 
- [x] Core: needs new YAML writer - #6374

New features:
- [x] Add `version` field to the YAML files with training data format version - #6246
- [x] Converter tool from `1.0` to `2.0` - #6402

### LINKED ISSUE #6246: Add version string to the 2.0 training data format
Part of #6129

The 2.0 YAML training data format files (NLU, Stories, Domain) should have a version string:
```yaml
version: '2.0'
```

Format of the version is: `major:minor`

- we start with `2.0` since there was a format before
- if no version was specified, we assume it's `2.0`

Rasa Open Source will print a warning message when:
- the training data can be updated to a later version (Rasa Open Source will try to read the files)
- the training data version if ahead of the used Rasa Open Source version (these files will be skipped)

### PR DESCRIPTION
**Pull request by degiz**
Closes #6246

**Proposed changes**:
- Users are warned if training data files are not compatible
- If no version is specified, latest one is assumed

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/validator.py [hunk 1]
```diff
@@ -257,3 +267,48 @@ def verify_domain_validity(self) -> bool:
...
+            return True
+
+        if version.parse(LATEST_TRAINING_DATA_FORMAT_VERSION) >= version.parse(
+            version_string
+        ):
+            return True
+        else:
+            raise_warning(
+                f"Training data file {filename} has a greater format version that "
+                f"your Rasa Open Source: "
```
Comment (federicotdn) [inline rasa/validator.py:308]:
```suggestion
                f"Training data file {filename} has a greater format version than "
                f"your Rasa Open Source installation: "
```

File: rasa/validator.py [hunk 2]
```diff
@@ -257,3 +267,48 @@ def verify_domain_validity(self) -> bool:
...
+                f"version '{LATEST_TRAINING_DATA_FORMAT_VERSION}' file.",
+                docs=DOCS_BASE_URL,
+            )
+            return True
+
+        if version.parse(LATEST_TRAINING_DATA_FORMAT_VERSION) >= version.parse(
+            version_string
+        ):
+            return True
+        else:
```
Comment (federicotdn) [inline rasa/validator.py:305]: Optional: remove `else` and de-indent code from here onwards.

File: rasa/validator.py [hunk 3]
```diff
@@ -257,3 +267,48 @@ def verify_domain_validity(self) -> bool:
...
+        if not isinstance(yaml_file_content, dict):
+            raise ValueError(f"Failed to validate {filename}.")
+
+        version_string = yaml_file_content.get(KEY_TRAINING_DATA_FORMAT_VERSION, "")
+        if not version_string:
+            raise_warning(
+                f"Training data file {filename} doesn't have a "
+                f"'{KEY_TRAINING_DATA_FORMAT_VERSION}' key. "
+                f"Rasa Open Source will read the file as "
+                f"version '{LATEST_TRAINING_DATA_FORMAT_VERSION}' file.",
```
Comment (federicotdn) [inline rasa/validator.py:296]:
```suggestion
                f"a version '{LATEST_TRAINING_DATA_FORMAT_VERSION}' file.",
```

File: rasa/validator.py [hunk 4]
```diff
@@ -257,3 +267,48 @@ def verify_domain_validity(self) -> bool:
...
+            filename: Name of the validated file.
+
+        Returns:
+            `True` if the file can be processed by current version of Rasa Open Source,
+            `False` otherwise.
+        """
+        if not isinstance(yaml_file_content, dict):
+            raise ValueError(f"Failed to validate {filename}.")
+
+        version_string = yaml_file_content.get(KEY_TRAINING_DATA_FORMAT_VERSION, "")
```
Comment (federicotdn) [inline rasa/validator.py:290]:
```suggestion
        version_string = yaml_file_content.get(KEY_TRAINING_DATA_FORMAT_VERSION)
```

**Comment by degiz** [comment]
Thanks a lot for catching that! Added some code and tests for this case.

**Pull request merged**